src/cuda/backend.f90

```diff
@@ -0,0 +1,404 @@
+                                  self%ythreads, self%yblocks)
+
+   end subroutine transeq_y_cuda
+
+   subroutine transeq_z_cuda(self, du, dv, dw, u, v, w, dirps)
+      implicit none
+
+      class(cuda_backend_t) :: self
+      class(field_t), intent(inout) :: du, dv, dw
+      class(field_t), intent(in) :: u, v, w
+      type(dirps_t), intent(in) :: dirps
+
+      print*, 'transeq_z_cuda'
+      ! w, u, v is reordered so that we pass w, u, v
+      call self%transeq_cuda_dist(dw, du, dv, w, u, v, dirps, &
+                                  self%zthreads, self%zblocks)
+
+   end subroutine transeq_z_cuda
+
+   subroutine transeq_cuda_dist(self, du, duu, d2u, u, v, w, dirps, &
+                                blocks, threads)
+      implicit none
+
+      class(cuda_backend_t) :: self
+      class(field_t), intent(inout) :: du, duu, d2u
+      class(field_t), intent(in) :: u, v, w
+      type(dirps_t), intent(in) :: dirps
+      type(dim3), intent(in) :: blocks, threads
+
+      class(field_t), pointer :: temp_du, temp_duu, temp_d2u, &
+                                 temp_dv, temp_dvu, temp_d2v, &
+                                 temp_dw, temp_dwu, temp_d2w
+
+      real, device, pointer, dimension(:, :, :) :: du_dev, duu_dev, d2u_dev, &
+                                                   dv_dev, dvu_dev, d2v_dev, &
+                                                   dw_dev, dwu_dev, d2w_dev
+
+      print*, 'transeq_cuda_dist'
+      ! MPI communication for halo data
+      ! first slice the halo data
+      !call slice_layers<<<blocks, threads>>>(u, buff_send_u_b, buff_send_u_e, derps%n_halo)
+      !call slice_layers<<<blocks, threads>>>(v, buff_send_v_b, buff_send_v_e, derps%n_halo)
+      !call slice_layers<<<blocks, threads>>>(w, buff_send_w_b, buff_send_w_e, derps%n_halo)
+
+      ! then send/recv halos
+      !call communicate_sendrecv(
+      !   buff_send_u_b, buff_send_u_e, buff_recv_u_b, buff_recv_u_e, &
+      !   derps%n_halo*derps%n_perp, derps%prev_rank, derps%next_rank, self%MPI_FP_PREC &
+      !)
+      !call communicate_sendrecv(
+      !   buff_send_v_b, buff_send_v_e, buff_recv_v_b, buff_recv_v_e, &
+      !   derps%n_halo*derps%n_perp, derps%prev_rank, derps%next_rank, self%MPI_FP_PREC &
+      !)
+      !call communicate_sendrecv(
+      !   buff_send_w_b, buff_send_w_e, buff_recv_w_b, buff_recv_w_e, &
+      !   derps%n_halo*derps%n_perp, derps%prev_rank, derps%next_rank, self%MPI_FP_PREC &
+      !)
+
+      ! distder_cuda
+
+      ! get some fields for storing the result
+      temp_du => self%allocator%get_block()
+      temp_duu => self%allocator%get_block()
+      temp_d2u => self%allocator%get_block()
+
+      select type(temp_du)
+      type is (cuda_field_t); du_dev => temp_du%data_d
+      end select
+      select type(temp_duu)
+      type is (cuda_field_t); duu_dev => temp_duu%data_d
+      end select
+      select type(temp_d2u)
+      type is (cuda_field_t); d2u_dev => temp_d2u%data_d
+      end select
```

Comment:
We obtain a device array pointer here and after this stage we can call kernels just like in the tests